x/streamswap/README.md

```diff
@@ -0,0 +1,121 @@
+
+The first two mechanisms are not well suited for early stage startups, where
+the token sale price is usually defined by a founding team and can't be
+impacted by the ecosystem wisdom. False marketing actions are usually setup
+to support their initial price.
+
+The latter mechanism is not democratic - big entities can control the
+price movements or place big orders leaving smaller investors with nothing.
+
+## Design
+
+### Sale Creation
+
+[Sale](https://github.com/osmosis-labs/osmosis/blob/main/proto/osmosis/streamswap/v1/state.proto#L11) object represent a particular token sale event and describes the main
+required parameters guarding the sale process:
+
+- `treasury`: address where the sale earnings will go. When the sale is over,
+  anyone can trigger a [`MsgFinalizeSale`](https://github.com/osmosis-labs/osmosis/blob/main/proto/osmosis/streamswap/v1/tx.proto#L42)
+  to clean up the sale state and move the earning to the treasury.
+- `id`: unique identifier of the sale.
+- `token_out`: denom to sale (distributed to the investors).
+  Also known as a base currency.
+- `token_in`: payment denom - used to buy `token_out`.
+  Also known as quote currency.
+- `token_out_supply`: total initial supply of `token_in` to sale.
+- `start_time`: Unix timestamp when the sale starts.
+- `end_time`: Unix timestamp when the sale ends.
+- `name`: Name of the sale.
+- `url`: an external resource describing a sale. Can be IPFS link or a
+  commonwealth post.
+
+The `Sale` object contains also other internal attributes to describe the current
+status of the sale.
+
+Anyone can create a `Sale` by sending [`MsgCreateSale`](https://github.com/osmosis-labs/osmosis/blob/robert%2Fstreamswap-spec/proto/osmosis/streamswap/v1/tx.proto#L21) transaction.
+When doing so, `token_out` amount of `Sale.token_out` tokens will be debited from
+his account and escrowed in the module account to distribute to Sale investors.
+Moreover the creator will be charged `sale_creation_fee` (module param) and the
+fee will be transferred to `sale_creation_fee_recipient` (module param) account.
+This fee is not recoverable.
+
+See other [module parameters](https://github.com/osmosis-labs/osmosis/main/proto/osmosis/streamswap/v1/params.proto) which control the sale creation.
+
+### Investing and distribution mechanism
+
+Anyone can join a sale by sending a [MsgSubscribe](https://github.com/osmosis-labs/osmosis/blob/main/proto/osmosis/streamswap/v1/tx.proto#L13) transaction.
+When doing so, the transaction author has to specify the `amount` he wants to spend in the sale.
+That `amount` will be credited from tx author and pledged to the sale.
+
+`MsgSubscribe` can be submitted at any time after the sale is created and before it's end time.
+
+From that moment, the investor will join the **token sale distribution stream**:
+
+- The distribution happens discretely in rounds. Each round is 1 second long.
+  We define: `total_rounds := (sale.end_time - sale.start_time) / round_duration`.
```

Comment:
Why even divide if `round_duration` is always going to be 1? Also is there a reason why each round is always 1?

Reply:
Well, this is the definition. Currently round_duration is set to 1second, but that can change.